### ISSUE TITLE
Possible bug with `isinstance` type-narrowing

### ISSUE THREAD

**Issue by tylerlaprade** [microsoft/pyright#9441]
Hi, I'm not sure if this is a bug in the pre-release version or if it's working as intended. `Model` in this case comes from Django, and its metaclass is `ModelBase`. This code typechecks just fine on the release version but errors on the commented line when on the pre-release version. The detected type is `model_or_iterable: list[type[Model]] | type[Model] | Iterable[type[Model]]`.
```python3
class Model(AltersData, metaclass=ModelBase):
```
```python3
class AbacusAdminSite(admin.AdminSite):
    """Custom admin site class for some UI tweaks."""

    @override
    def register(
        self,
        model_or_iterable: 'type[Model] | Iterable[type[Model]]',
        admin_class: type['ModelAdmin[Any]'] | None = None,
        **options,
    ) -> None:
        super().register(model_or_iterable, admin_class=admin_class, **options)
        if isinstance(model_or_iterable, ModelBase):
            model_or_iterable = [model_or_iterable]
        for model in model_or_iterable:  // "type[Model]" is not iterable
  "__iter__" method not definedPylance[reportGeneralTypeIssues]
            if not _should_make_proxy_model(model):
                continue
            soft_deleted_model = create_proxy_model(model)
            soft_deleted_model_admin = (
                create_softdeleted_model_admin(admin_class) if admin_class else None
            )
            super().register(
                soft_deleted_model, admin_class=soft_deleted_model_admin, **options
            )
```

**Comment by erictraut** [comment]
Can you provide a minimal self-contained example? The code posted above contains many symbols that are not defined.

**Comment by tylerlaprade** [comment]
```python3
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterable


class ModelBase(type):
    pass
class Model(metaclass=ModelBase):
    pass
def register(
        model_or_iterable: 'type[Model] | Iterable[type[Model]]',
    ):
        if isinstance(model_or_iterable, ModelBase):
            model_or_iterable = [model_or_iterable]
        for _model in model_or_iterable:   # "type[Model]" is not iterable "__iter__" method not definedPylance[reportGeneralTypeIssues]
            pass

```

**Comment by erictraut** [comment]
You mentioned the "pre-release version". Do you suspect that this is a regression? Are you building your own version of pyright from source? I'm not seeing any difference in behavior between the main branch and the latest published version of pyright.

I'm not sure I'd classify this as a bug, but I agree there's an opportunity for further narrowing in the negative case when using a metaclass in an `isinstance` type guard.

```python
class Meta(type):
    pass

class A(metaclass=Meta):
    pass

def func(m: type[A] | str):
    if not isinstance(m, Meta):
        reveal_type(m)  # type[A] | str, but could be just str
```

**Comment by tylerlaprade** [comment]
Perfect, thanks Eric! Sorry for the confusion.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.391.
